File: back/src/controller/usuario/pasosReceta.controller.js
const db = require('../../db/conection');

const getPasosRecetas = async (req, res) => {
  const usuarioId = req.session.id;
  if (!usuarioId) {
    return res.status(401).json({ error: 'No estás autenticado' });
  }

  const sql = 'SELECT * FROM pasos_por_receta WHERE usuario_id = ?';
  try {
    const [pasos] = await db.query(sql, [usuarioId]);
    if (pasos.length === 0) {
      return res.status(404).json({ mensaje: 'No se encontraron pasos para este usuario' });
    }
    return res.status(200).json(pasos);
  } catch (error) {
    console.error(error);
    return res.status(500).json({ error: 'Error interno del servidor' });
  }
};

const getUnPasoReceta = async (req, res) => {
  const { id } = req.params;
  const usuarioId = req.session.id;
  if (!usuarioId) {
    return res.status(401).json({ error: 'No estás autenticado' });
  }
  if (!id) {
    return res.status(400).json({ error: 'Falta el ID del paso' });
  }
  
  const sql = 'SELECT * FROM pasos_por_receta WHERE id = ? AND usuario_id = ?';
  try {
    const [paso] = await db.query(sql, [id, usuarioId]);
    if (paso.length === 0) {
      return res.status(404).json({ error: 'Paso no encontrado o no tienes permiso para verlo' });
    }
    return res.status(200).json(paso[0]);
  } catch (error) {
    console.error(error);
    return res.status(500).json({ error: 'Error interno del servidor' });
  }
};

const postPasoReceta = async (req, res) => {
  const { recetaId, elaboracion, imagen } = req.body;
  const usuarioId = req.session.id;
  if (!usuarioId) {
    return res.status(401).json({ error: 'No estás autenticado' });
  }
  if (!recetaId || !elaboracion || !imagen) {
    return res.status(400).json({ error: 'Faltan datos obligatorios' });
  }

  const sql = 'INSERT INTO pasos_por_receta (receta_id, elaboracion, imagen, usuario_id) VALUES (?, ?, ?, ?)';
  try {
    const [result] = await db.query(sql, [recetaId, elaboracion, imagen, usuarioId]);
    return res.status(201).json({ mensaje: 'Paso agregado correctamente', id: result.insertId });
  } catch (error) {
    console.error(error);
    return res.status(500).json({ error: 'Error interno del servidor' });
  }
};

const putPasoReceta = async (req, res) => {
  const { id } = req.params;
  const { recetaId, elaboracion, imagen } = req.body;
  const usuarioId = req.session.id;

  if (!usuarioId) {
    return res.status(401).json({ error: 'No estás autenticado' });
  }
  if (!id || !recetaId || !elaboracion || !imagen) {
    return res.status(400).json({ error: 'Faltan datos obligatorios' });
  }

  const sql = 'UPDATE pasos_por_receta SET receta_id = ?, elaboracion = ?, imagen = ? WHERE id = ? AND usuario_id = ?';
  try {
    const [result] = await db.query(sql, [recetaId, elaboracion, imagen, id, usuarioId]);
    if (result.affectedRows === 0) {
      return res.status(404).json({ error: 'Paso no encontrado o no tienes permiso para editarlo' });
    }
    return res.status(200).json({ mensaje: 'Paso actualizado correctamente' });
  } catch (error) {
    console.error(error);
    return res.status(500).json({ error: 'Error interno del servidor' });
  }
};

const deletePasoReceta = async (req, res) => {
  const { id } = req.params;
  const usuarioId = req.session.id;
  if (!usuarioId) {
    return res.status(401).json({ error: 'No estás autenticado' });
  }
  if (!id) {
    return res.status(400).json({ error: 'Falta el ID del paso' });
  }

  const sql = 'DELETE FROM pasos_por_receta WHERE id = ? AND usuario_id = ?';
  try {
    const [result] = await db.query(sql, [id, usuarioId]);
    if (result.affectedRows === 0) {
      return res.status(404).json({ error: 'Paso no encontrado o no tienes permiso para eliminarlo' });
    }
    return res.status(200).json({ mensaje: 'Paso eliminado correctamente' });
  } catch (error) {
    console.error(error);
    return res.status(500).json({ error: 'Error interno del servidor' });
  }
};

module.exports = {
  getPasosReceta,
  getUnPasoReceta,
  postPasoReceta,
  putPasoReceta,
  deletePasoReceta
};

File: ListaDeLaCompra_copia/back_py/controller/pasos_receta_controller.py
from db.conexion import get_connection
from flask import session

def get_pasos_receta_por_usuario():
    usuario_id = session.get('usuario_id')
    if not usuario_id:
        return {'error': 'No estás autenticado'}, 401
    sql = "SELECT * FROM pasos_por_receta WHERE usuario_id = %s"
    try:
        db = get_connection()
        cursor = db.cursor(dictionary=True)
        cursor.execute(sql, (usuario_id,))
        pasos_list = cursor.fetchall()
        if not pasos_list:
            return {'mensaje': 'No se encontraron pasos para este usuario'}, 404
        return pasos_list, 200
    except Exception:
        return {'error': 'Error interno del servidor'}, 500

def get_pasos_receta_por_receta(receta_id):
    usuario_id = session.get('usuario_id')
    if not usuario_id:
        return {'error': 'No estás autenticado'}, 401
    if not receta_id:
        return {'error': 'Falta el ID de la receta'}, 400
    sql = "SELECT * FROM pasos_por_receta WHERE receta_id = %s AND usuario_id = %s"
    try:
        db = get_connection()
        cursor = db.cursor(dictionary=True)
        cursor.execute(sql, (receta_id, usuario_id))
        pasos_list = cursor.fetchall()
        if not pasos_list:
            return {'mensaje': 'No se encontraron pasos para esta receta'}, 404
        return pasos_list, 200
    except Exception:
        return {'error': 'Error interno del servidor'}, 500

def get_un_paso_receta(id):
    usuario_id = session.get('usuario_id')
    if not usuario_id:
        return {'error': 'No estás autenticado'}, 401
    if not id:
        return {'error': 'Falta el ID del paso'}, 400
    sql = "SELECT * FROM pasos_por_receta WHERE id = %s AND usuario_id = %s"
    try:
        db = get_connection()
        cursor = db.cursor(dictionary=True)
        cursor.execute(sql, (id, usuario_id))
        paso_dict = cursor.fetchone()
        if not paso_dict:
            return {'error': 'Paso no encontrado o no tienes permiso para verlo'}, 404
        return paso_dict, 200
    except Exception:
        return {'error': 'Error interno del servidor'}, 500

def post_paso_receta(receta_id, elaboracion, imagen):
    usuario_id = session.get('usuario_id')
    if not usuario_id:
        return {'error': 'No estás autenticado'}, 401
    if not receta_id or not elaboracion:
        return {'error': 'Faltan datos obligatorios'}, 400
    sql = "INSERT INTO pasos_por_receta (receta_id, elaboracion, imagen, usuario_id) VALUES (%s, %s, %s, %s)"
    try:
        db = get_connection()
        cursor = db.cursor()
        cursor.execute(sql, (receta_id, elaboracion, imagen, usuario_id))
        db.commit()
        return {'mensaje': 'Paso agregado correctamente'}, 201
    except Exception:
        db.rollback()
        return {'error': 'Error interno del servidor'}, 500

def post_varios_pasos_receta(pasos):
    usuario_id = session.get('usuario_id')
    if not usuario_id:
        return {'error': 'No estás autenticado'}, 401
    if not isinstance(pasos, list) or len(pasos) == 0:
        return {'error': 'Faltan pasos para agregar'}, 400
    sql = "INSERT INTO pasos_por_receta (receta_id, elaboracion, imagen, usuario_id) VALUES (%s, %s, %s, %s)"
    db = get_connection()
    cursor = db.cursor()
    try:
        db.start_transaction()
        for paso in pasos:
            receta_id = paso.get('receta_id')
            elaboracion = paso.get('elaboracion')
            imagen = paso.get('imagen')
            cursor.execute(sql, (receta_id, elaboracion, imagen, usuario_id))
        db.commit()
        return {'mensaje': 'Pasos agregados correctamente'}, 201
    except Exception:
        db.rollback()
        return {'error': 'Error interno del servidor'}, 500

def put_paso_receta(id, receta_id, elaboracion, imagen):
    usuario_id = session.get('usuario_id')
    print(usuario_id)
    if not usuario_id:
        return {'error': 'No estás autenticado'}, 401
    print(id, receta_id, elaboracion, imagen)
    if not id or not receta_id:
        return {'error': 'Faltan datos obligatorios'}, 400
    sql = "UPDATE pasos_por_receta SET receta_id = %s, elaboracion = %s, imagen = %s WHERE id = %s AND usuario_id = %s"
    try:
        db = get_connection()
        cursor = db.cursor()
        cursor.execute(sql, (receta_id, elaboracion, imagen, id, usuario_id))
        db.commit()
        if cursor.rowcount == 0:
            return {'error': 'Paso no encontrado o no tienes permiso para editarlo'}, 404
        return {'mensaje': 'Paso actualizado correctamente'}, 200
    except Exception:
        db.rollback()
        return {'error': 'Error interno del servidor'}, 500

def delete_paso_receta(id):
    usuario_id = session.get('usuario_id')
    if not usuario_id:
        return {'error': 'No estás autenticado'}, 401
    if not id:
        return {'error': 'Falta el ID del paso'}, 400
    sql = "DELETE FROM pasos_por_receta WHERE id = %s AND usuario_id = %s"
    try:
        db = get_connection()
        cursor = db.cursor()
        cursor.execute(sql, (id, usuario_id))
        db.commit()
        if cursor.rowcount == 0:
            return {'error': 'Paso no encontrado o no tienes permiso para eliminarlo'}, 404
        return {'mensaje': 'Paso eliminado correctamente'}, 200
    except Exception:
        db.rollback()
        return {'error': 'Error interno del servidor'}, 500
